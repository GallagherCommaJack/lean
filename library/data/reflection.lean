import data.list
open sum

namespace lean
  definition name := list (string + num)

  namespace syntax
    inductive level : Type₁ :=
    | meta   : name → level
    | zero   : level
    | succ   : level → level
    | max    : level → level → level
    | imax   : level → level → level
    | param  : name → level
    | global : name → level

    inductive expr : Type₁ :=
    | var   : ℕ → expr
    | const : name  → list level → expr
    | loc   : name  → expr → expr
    | meta  : name  → expr → expr
    | sort  : level → expr
    | pi    : name  → expr → expr → expr
    | lam   : name  → expr → expr → expr
    | app   : expr  → expr → expr
    | elet  : name  → expr → expr → expr → expr
  end syntax
end lean
