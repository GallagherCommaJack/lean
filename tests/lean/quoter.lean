import data.reflection

eval quote (λ x, x) -- lambdas, variables, metavariables
eval quote (λ x, nat.succ x)
eval quote ((λ x, x) nat.zero)
