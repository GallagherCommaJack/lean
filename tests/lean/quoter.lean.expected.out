lam
  (meta [inr 0, inr 18]
     (sort
        (param
           [inl
             (str (char.mk ff tt tt ff tt tt ff ff)
                (str (char.mk ff tt ff tt tt tt tt tt) (str (char.mk ff ff tt tt ff ff ff tt) empty)))])))
  (var 0)
lam
  (sort
     (param
        [inl
          (str (char.mk ff tt tt ff tt tt ff ff)
             (str (char.mk ff tt ff tt tt tt tt tt) (str (char.mk ff ff tt tt ff ff ff tt) empty)))]))
  (lam (var 0) (var 0))
lam
  (const
     [inl
       (str (char.mk ff tt tt ff tt tt tt ff)
          (str (char.mk ff tt tt ff ff ff ff tt) (str (char.mk ff tt tt tt ff tt ff ff) empty)))]
     nil)
  (app
     (const
        [inl
          (str (char.mk ff tt tt tt ff ff tt tt)
             (str (char.mk ff tt tt tt ff tt ff tt)
                (str (char.mk ff tt tt ff ff ff tt tt) (str (char.mk ff tt tt ff ff ff tt tt) empty)))), inl
          (str (char.mk ff tt tt ff tt tt tt ff)
             (str (char.mk ff tt tt ff ff ff ff tt) (str (char.mk ff tt tt tt ff tt ff ff) empty)))]
        nil)
     (var 0))
app
  (lam
     (const
        [inl
          (str (char.mk ff tt tt ff tt tt tt ff)
             (str (char.mk ff tt tt ff ff ff ff tt) (str (char.mk ff tt tt tt ff tt ff ff) empty)))]
        nil)
     (var 0))
  (const
     [inl
       (str (char.mk ff tt tt tt tt ff tt ff)
          (str (char.mk ff tt tt ff ff tt ff tt)
             (str (char.mk ff tt tt tt ff ff tt ff) (str (char.mk ff tt tt ff tt tt tt tt) empty)))), inl
       (str (char.mk ff tt tt ff tt tt tt ff)
          (str (char.mk ff tt tt ff ff ff ff tt) (str (char.mk ff tt tt tt ff tt ff ff) empty)))]
     nil)
